machine_learning/automatic_differentiation.py

```
@@ -0,0 +1,324 @@
+                else:
+                    partial_deriv[param] = dparam_dtarget
+
+                if param.result_of:
+                    if param.result_of != OpType.NOOP:
+                        operation_queue.append(param.result_of)
+
+        if source in partial_deriv:
+            return partial_deriv[source]
+        return None
+
+    def derivative(self, param: Variable, operation: Operation) -> np.ndarray:
+        """
+        Compute the derivative of given operation/function
+
+        Args:
+            param: variable to be differentiated
+            operation: function performed on the input variable
+
+        Returns:
+            Derivative of input variable with respect to the output of
+            the operation
+        """
+        params = operation.params
+
+        derivative = None
+
+        if operation == OpType.ADD:
+            derivative = np.ones_like(params[0].numpy(), dtype=np.float64)
+        elif operation == OpType.SUB:
+            if params[0] == param:
+                derivative = np.ones_like(params[0].numpy(), dtype=np.float64)
+            else:
+                derivative = -np.ones_like(params[1].numpy(), dtype=np.float64)
+        elif operation == OpType.MUL:
+            derivative = (
+                params[1].numpy().T if params[0] == param else params[0].numpy().T
+            )
+        elif operation == OpType.DIV:
+            if params[0] == param:
+                derivative = 1 / params[1].numpy()
+            else:
+                derivative = -params[0].numpy() / (params[1].numpy() ** 2)
+        elif operation == OpType.MATMUL:
+            derivative = (
+                params[1].numpy().T if params[0] == param else params[0].numpy().T
+            )
+        elif operation == OpType.POWER:
+            power = operation.other_params["power"]
+            derivative = power * (params[0].numpy() ** (power - 1))
+        return derivative
```

Do we want to `raise ValueError("Invalid operation")` here?